### DISCUSSION TITLE
[Proposal]: async method exception filter

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#9000] [Language Ideas]
# async method exception filter

Champion issue: https://github.com/dotnet/csharplang/issues/9001

## Summary
[summary]: #summary

Allow customizing the exception handling produced for async methods.  
Adding the following attribute:  
`[module: AsyncExceptionFilter(typeof(MyExceptionHandler)]`
would result in the generated `try/catch` having a modified `catch` clause:  
```csharp
catch (Exception e) when (MyExceptionHandler.Report(e); true)
{
    ... existing handling logic ...
}
```
where we could find the following method `void Report(Exception)`.

## Motivation
[motivation]: #motivation

There's exception handling that is better done inside the async method so that it can see the original stacktrace and can produce useful crash dumps, but it can be painful to repeat in every async method.  
The roslyn IDE code is an example, where there is desire to add the following pattern to every async method:

```csharp
private async Task ExampleAsync()
{
    try
    {
        ... async code ...
    }
    catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e))
    {
        throw ExceptionUtilities.Unreachable;
    }
}
```

[Example](http://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationWithAnalyzers.cs,581)
[FatalError](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/FatalError.cs,3ac0310b7374dbc1,references)

## Design Meetings

- https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-09-28.md#ungrouped

**Comment by mburbea** [comment]
This is something I'd love to have, but I'm curious will this retain the contextual variables? 
e.g.
```cs
public Task M()
{
    using (_logger.BeginScope("label"))
             await SomeMethodThatThrows();
}
```

If I wrap this with a logging filter the scope will be lost normally in the async transformation.

```cs
public Task M()
{
    using (_logger.BeginScope("label"))
    {
       try{  await SomeMethodThatThrows();}
       catch(Exception e) when (LoggingFilter(_logger, e)){}
    }
}
```

**Comment by HaloFour** [comment]
Would it be useful to also be able to apply this attribute to specific methods to opt-in to the exception filter where needed instead of it affecting the entire assembly?
